### PR TITLE
one policy has many rules

### PR DESCRIPTION
**Pull request by jikui**

### CODE REVIEW

File: app/waf/src/controllers/application.controller.ts [hunk 1]
```diff
@@ -257,24 +257,23 @@ export class ApplicationController {
...
           rule.conditions = await this.ruleRepository
             .conditions(rule.id)
             .find();
           rule.actions = await this.ruleRepository.actions(rule.id).find();
         }
-        params.wafs = await this.wafpolicyRepository.find();
       }
     }
 
+    params.wafs = await this.wafpolicyRepository.find();
```
Comment (zongzw) [inline app/waf/src/controllers/application.controller.ts]: Will we attach all wafpolicy to the deployment without filter condition?

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 2]
```diff
@@ -37,9 +36,6 @@ export class EndpointpolicyController {
   async create(
     @requestBody() endpointpolicy: Partial<Endpointpolicy>,
   ): Promise<Endpointpolicy> {
-    if (!endpointpolicy.id) {
-      endpointpolicy.id = uuid();
-    }
     return await this.endpointpolicyRepository.create(endpointpolicy);
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: If no id input from API, the uuid is not generated here, right? Did you test that?

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 3]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
-    @param.path.string('id') id: string,
-    @requestBody() endpointpolicy: Partial<Endpointpolicy>,
-  ): Promise<void> {
-    await this.endpointpolicyRepository.replaceById(id, endpointpolicy);
+  async deleteById(@param.path.string('id') id: string): Promise<void> {
+    await this.endpointpolicyRepository.deleteById(id);
   }
 
-  @del(prefix + '/endpointpolicies/{id}', {
+  @post(prefix + '/endpointpolicies/{endpointpolicy_id}/rules', {
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

This parameter name will be exposed via API spec, so we need to make its naming be consistent with others.

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 4]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+    responses: {
+      '200': {
+        description: 'Rules add to rule success',
+        content: {'application/json': {schema: {'x-ts-type': Rule}}},
+      },
+    },
+  })
+  async createEndpointpolicyRule(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 5]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+    @requestBody() rule: Partial<Rule>,
+  ): Promise<Rule> {
+    return await this.endpointpolicyRepository
+      .rules(endpointpolicyId)
+      .create(rule);
+  }
+
+  @get(prefix + '/endpointpolicies/{endpointpolicy_id}/rules', {
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 6]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+        content: {
+          'application/json': {
+            schema: {type: 'array', items: {'x-ts-type': Rule}},
+          },
+        },
+      },
+    },
+  })
+  async getRules(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 7]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+  @get(prefix + '/endpointpolicies/{endpointpolicy_id}/rules/{rule_id}', {
+    responses: {
+      '200': {
+        description: 'Rule model instance',
+        content: {'application/json': {schema: {'x-ts-type': Rule}}},
+      },
+    },
+  })
+  async getRuleByID(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 8]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+        'Condition ' +
+          ruleId +
+          ' for ruleId ' +
+          endpointpolicyId +
+          ' not found.',
+      );
+    }
+  }
+
+  @del(prefix + '/endpointpolicies/{endpointpolicy_id}/rules/{rule_id}', {
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 9]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
       '204': {
-        description: 'Endpointpolicy DELETE success',
+        description: 'Rules DELETE success',
       },
     },
   })
-  async deleteById(@param.path.string('id') id: string): Promise<void> {
-    await this.endpointpolicyRepository.deleteById(id);
+  async deleteRuleByID(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 10]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+    @param.path.string('rule_id') ruleId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: rule_id --> ruleId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 11]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+    responses: {
+      '200': {
+        description: 'Rule model instance',
+        content: {'application/json': {schema: {'x-ts-type': Rule}}},
+      },
+    },
+  })
+  async updateRuleByID(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
+    @param.path.string('rule_id') ruleId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: rule_id --> ruleId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 12]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+  @patch(prefix + '/endpointpolicies/{endpointpolicy_id}/rules/{rule_id}', {
+    responses: {
+      '200': {
+        description: 'Rule model instance',
+        content: {'application/json': {schema: {'x-ts-type': Rule}}},
+      },
+    },
+  })
+  async updateRuleByID(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId

File: app/waf/src/controllers/endpointpolicy.controller.ts [hunk 13]
```diff
@@ -119,28 +99,110 @@ export class EndpointpolicyController {
...
+  async deleteRuleByID(
+    @param.path.string('endpointpolicy_id') endpointpolicyId: string,
+    @param.path.string('rule_id') ruleId: string,
+  ) {
+    await this.endpointpolicyRepository
+      .rules(endpointpolicyId)
+      .delete({id: ruleId});
+  }
+
+  @patch(prefix + '/endpointpolicies/{endpointpolicy_id}/rules/{rule_id}', {
```
Comment (zhaoqin-github) [inline app/waf/src/controllers/endpointpolicy.controller.ts]: endpointpolicy_id --> endpointpolicyId
rule_id --> ruleId

File: app/waf/src/controllers/application.controller.ts (hunk 1 above)
Comment (jikui) [reply app/waf/src/controllers/application.controller.ts]: for this deploy functionality . we could leave it alone.

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 2 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: will check again.

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 5 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: acccepted

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 6 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: accepted

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 9 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: accepted

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 10 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: accepted it

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 13 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: accepted

File: app/waf/src/controllers/endpointpolicy.controller.ts (hunk 2 above)
Comment (jikui) [reply app/waf/src/controllers/endpointpolicy.controller.ts]: pick it back.

**Pull request merged**